Build does not work (Quarkus 3.11.1) :  java.lang.NoSuchFieldError: type at io.quarkiverse.kerberos.deployment.KerberosBuildStep.register2

Hello,

I just cannot use the extension as I get a build error when starting my application.
I'm using quarkus 3.11.1. The project has been generated from code.quarkus.io by selecting Rest and Kerberos extensions.

> [INFO] Scanning for projects...
> [INFO] 
> [INFO] --------------< com.inetum.bd.kerberos:kerberos-test-app >--------------
> [INFO] Building kerberos-test-app 1.0.0-SNAPSHOT
> [INFO]   from pom.xml
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO] 
> [INFO] --- quarkus:3.11.1:dev (default-cli) @ kerberos-test-app ---
> [INFO] Invoking resources:3.3.1:resources (default-resources) @ kerberos-test-app
> [INFO] Copying 1 resource from src\main\resources to target\classes
> [INFO] Invoking quarkus:3.11.1:generate-code (default) @ kerberos-test-app
> [INFO] Invoking compiler:3.12.1:compile (default-compile) @ kerberos-test-app
> [INFO] Recompiling the module because of added or removed source files.
> [INFO] Compiling 3 source files with javac [debug release 17] to target\classes
> [INFO] Invoking resources:3.3.1:testResources (default-testResources) @ kerberos-test-app
> [INFO] skip non existing resourceDirectory C:\dev\git\kerberos-test-app\src\test\resources
> [INFO] Invoking quarkus:3.11.1:generate-code-tests (default) @ kerberos-test-app
> [INFO] Invoking compiler:3.12.1:testCompile (default-testCompile) @ kerberos-test-app
> [INFO] Recompiling the module because of changed dependency.
> [INFO] Compiling 2 source files with javac [debug release 17] to target\test-classes
> Listening for transport dt_socket at address: 5005
> Press [e] to edit command line args (currently ''), [h] for more options>
> 2024-06-07 12:32:28,211 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
> 2024-06-07 12:32:28,589 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
> 	[error]: Build step io.quarkiverse.kerberos.deployment.KerberosBuildStep#register2 threw an exception: java.lang.NoSuchFieldError: type
> 	at io.quarkiverse.kerberos.deployment.KerberosBuildStep.register2(KerberosBuildStep.java:59)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
> 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
> 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
> 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1495)
> 	at java.base/java.lang.Thread.run(Thread.java:833)
> 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
> 
> 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
> 	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:252)
> 	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:58)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:113)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:435)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:56)
> 	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
> 	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
> 	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
> 	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
> Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
> 	[error]: Build step io.quarkiverse.kerberos.deployment.KerberosBuildStep#register2 threw an exception: java.lang.NoSuchFieldError: type
> 	at io.quarkiverse.kerberos.deployment.KerberosBuildStep.register2(KerberosBuildStep.java:59)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
> 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
> 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
> 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1495)
> 	at java.base/java.lang.Thread.run(Thread.java:833)
> 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
> 
> 	at io.quarkus.builder.Execution.run(Execution.java:123)
> 	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
> 	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
> 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:327)
> 	... 9 more
> Caused by: java.lang.NoSuchFieldError: type
> 	at io.quarkiverse.kerberos.deployment.KerberosBuildStep.register2(KerberosBuildStep.java:59)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
> 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
> 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
> 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1495)
> 	at java.base/java.lang.Thread.run(Thread.java:833)
> 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)



pom.xml dependencies : 

`    <modelVersion>4.0.0</modelVersion>
    <groupId>com.inetum.bd.kerberos</groupId>
    <artifactId>kerberos-test-app</artifactId>
    <version>1.0.0-SNAPSHOT</version>

    <properties>
        <compiler-plugin.version>3.12.1</compiler-plugin.version>
        <maven.compiler.release>17</maven.compiler.release>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
        <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
        <quarkus.platform.version>3.11.1</quarkus.platform.version>
        <skipITs>true</skipITs>
        <surefire-plugin.version>3.2.5</surefire-plugin.version>
    </properties>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>${quarkus.platform.group-id}</groupId>
                <artifactId>${quarkus.platform.artifact-id}</artifactId>
                <version>${quarkus.platform.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-rest</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkiverse.kerberos</groupId>
            <artifactId>quarkus-kerberos</artifactId>
            <version>2.1.0</version>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-arc</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-junit5</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.rest-assured</groupId>
            <artifactId>rest-assured</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>`

--- a/deployment/pom.xml
+++ b/deployment/pom.xml
@@ -27,7 +27,7 @@
     </dependency>
     <dependency>
       <groupId>io.quarkus</groupId>
-      <artifactId>quarkus-resteasy-reactive-deployment</artifactId>
+      <artifactId>quarkus-rest-deployment</artifactId>
       <scope>test</scope>
     </dependency>
     <dependency>

--- a/deployment/src/main/java/io/quarkiverse/kerberos/deployment/KerberosBuildStep.java
+++ b/deployment/src/main/java/io/quarkiverse/kerberos/deployment/KerberosBuildStep.java
@@ -2,8 +2,6 @@ package io.quarkiverse.kerberos.deployment;
 
 import java.util.function.BooleanSupplier;
 
-import com.sun.security.auth.module.Krb5LoginModule;
-
 import io.quarkiverse.kerberos.runtime.KerberosAuthenticationMechanism;
 import io.quarkiverse.kerberos.runtime.KerberosIdentityProvider;
 import io.quarkiverse.kerberos.runtime.KerberosProducer;
@@ -15,7 +13,7 @@ import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBundleBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
-import io.quarkus.deployment.pkg.PackageConfig;
+import io.quarkus.deployment.pkg.NativeConfig;
 
 public class KerberosBuildStep {
 
@@ -36,7 +34,7 @@ public class KerberosBuildStep {
     @BuildStep(onlyIf = IsEnabled.class)
     public ReflectiveClassBuildItem reflection() {
         return ReflectiveClassBuildItem
-                .builder(Krb5LoginModule.class.getName(), "sun.security.jgss.krb5.Krb5MechFactory",
+                .builder("sun.security.jgss.krb5.Krb5MechFactory",
                         "sun.security.jgss.SunProvider", "sun.security.jgss.spnego.SpNegoMechFactory",
                         "javax.security.auth.kerberos.KerberosPrincipal", "javax.security.auth.kerberos.KerberosKey",
                         "javax.security.auth.kerberos.KeyTab", "javax.security.auth.kerberos.KerberosTicket")
@@ -49,14 +47,28 @@ public class KerberosBuildStep {
     }
 
     @BuildStep(onlyIf = IsEnabled.class)
+    ReflectiveClassBuildItem loginModules() {
+        return ReflectiveClassBuildItem
+                .builder(
+                        "com.sun.jmx.remote.security.FileLoginModule",
+                        "com.sun.security.auth.module.JndiLoginModule",
+                        "com.sun.security.auth.module.KeyStoreLoginModule",
+                        "com.sun.security.auth.module.Krb5LoginModule",
+                        "com.sun.security.auth.module.LdapLoginModule",
+                        "com.sun.security.auth.module.NTLoginModule",
+                        "com.sun.security.auth.module.UnixLoginModule")
+                .build();
+    }
+
+    @BuildStep(onlyIf = IsEnabled.class)
     public NativeImageResourceBundleBuildItem resourceBundleBuildItem() {
         return new NativeImageResourceBundleBuildItem("sun.security.util.AuthResources");
     }
 
     @BuildStep(onlyIf = IsEnabled.class)
     @Record(ExecutionTime.RUNTIME_INIT)
-    public void register2(KerberosRecorder rec, PackageConfig packageConfig) throws Exception {
-        if (packageConfig.type.equals(PackageConfig.NATIVE)) {
+    public void register2(KerberosRecorder rec, NativeConfig nativeConfig) throws Exception {
+        if (nativeConfig.enabled()) {
             rec.registerProviderForNative();
         }
     }

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -11,7 +11,7 @@
   <dependencies>
     <dependency>
       <groupId>io.quarkus</groupId>
-      <artifactId>quarkus-resteasy</artifactId>
+      <artifactId>quarkus-rest</artifactId>
     </dependency>
     <dependency>
       <groupId>io.quarkiverse.kerberos</groupId>
@@ -88,7 +88,7 @@
         </plugins>
       </build>
       <properties>
-        <quarkus.package.type>native</quarkus.package.type>
+        <quarkus.native.enabled>true</quarkus.native.enabled>
       </properties>
     </profile>
   </profiles>

--- a/integration-tests/src/main/java/io/quarkiverse/kerberos/it/FrontendResource.java
+++ b/integration-tests/src/main/java/io/quarkiverse/kerberos/it/FrontendResource.java
@@ -9,10 +9,12 @@ import jakarta.inject.Inject;
 import jakarta.ws.rs.GET;
 import jakarta.ws.rs.NotAuthorizedException;
 import jakarta.ws.rs.Path;
+import jakarta.ws.rs.WebApplicationException;
 import jakarta.ws.rs.core.HttpHeaders;
 
 import org.eclipse.microprofile.rest.client.inject.RestClient;
 import org.ietf.jgss.GSSContext;
+import org.jboss.resteasy.reactive.ClientWebApplicationException;
 
 import io.quarkiverse.kerberos.client.KerberosClientSupport;
 
@@ -52,7 +54,11 @@ public class FrontendResource {
     @GET
     @Path("without-kerberos-support")
     public String getIdentityWithoutKerberosSupport() {
-        return identityServiceClient.getIdentity();
+        try {
+            return identityServiceClient.getIdentity();
+        } catch (ClientWebApplicationException ex) {
+            throw new WebApplicationException(ex.getResponse().getStatus());
+        }
     }
 
     private class IdentityServiceAction implements PrivilegedExceptionAction<String> {
@@ -78,5 +84,5 @@ public class FrontendResource {
             }
             throw new RuntimeException("Kerberos ticket can not be created");
         }
-    };
+    }
 }

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
   <parent>
     <groupId>io.quarkiverse</groupId>
     <artifactId>quarkiverse-parent</artifactId>
-    <version>13</version>
+    <version>16</version>
   </parent>
   <groupId>io.quarkiverse.kerberos</groupId>
   <artifactId>quarkus-kerberos-parent</artifactId>
@@ -24,13 +24,7 @@
     <tag>HEAD</tag>
   </scm>
   <properties>
-    <compiler-plugin.version>3.8.1</compiler-plugin.version>
-    <maven.compiler.parameters>true</maven.compiler.parameters>
-    <maven.compiler.source>11</maven.compiler.source>
-    <maven.compiler.target>11</maven.compiler.target>
-    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-    <quarkus.version>3.1.0.Final</quarkus.version>
+    <quarkus.version>3.12.3</quarkus.version>
     <apacheds-all.version>2.0.0-M15</apacheds-all.version>
   </properties>
   <dependencyManagement>
@@ -71,10 +65,6 @@
           <groupId>io.quarkus</groupId>
           <artifactId>quarkus-maven-plugin</artifactId>
           <version>${quarkus.version}</version>
-        </plugin>
-        <plugin>
-          <artifactId>maven-compiler-plugin</artifactId>
-          <version>${compiler-plugin.version}</version>
         </plugin>
       </plugins>
     </pluginManagement>

--- a/runtime/pom.xml
+++ b/runtime/pom.xml
@@ -8,6 +8,7 @@
   </parent>
   <artifactId>quarkus-kerberos</artifactId>
   <name>Quarkus - Kerberos - Runtime</name>
+  <description>Add Kerberos authentication support to Quarkus for HTTP requests</description>
   <dependencies>
     <dependency>
       <groupId>io.quarkus</groupId>

--- a/runtime/src/main/java/io/quarkiverse/kerberos/runtime/KerberosIdentityProvider.java
+++ b/runtime/src/main/java/io/quarkiverse/kerberos/runtime/KerberosIdentityProvider.java
@@ -157,18 +157,18 @@ public class KerberosIdentityProvider implements IdentityProvider<NegotiateAuthe
                         throw new AuthenticationFailedException();
                     }
                 } catch (LoginException ex) {
-                    LOG.debugf("Login exception: %s", ex.getMessage());
+                    LOG.debugf(ex, "Login exception: %s", ex.getMessage());
                     throw new AuthenticationCompletionException(ex);
                 } catch (GSSException ex) {
-                    LOG.debugf("GSS exception: %s", ex.getMessage());
+                    LOG.debugf(ex, "GSS exception: %s", ex.getMessage());
                     throw new AuthenticationCompletionException(ex);
                 } catch (PrivilegedActionException ex) {
                     Throwable ex2 = ex.getCause() != null ? ex.getCause() : ex;
-                    LOG.debugf("PrivilegedAction failure: %s", ex2.getMessage());
+                    LOG.debugf(ex2, "PrivilegedAction failure: %s", ex2.getMessage());
                     throw new AuthenticationCompletionException(ex);
                 } catch (Throwable ex) {
                     Throwable ex2 = ex.getCause() != null ? ex.getCause() : ex;
-                    LOG.debugf("Authentication failure: %s", ex2.getMessage());
+                    LOG.debugf(ex2, "Authentication failure: %s", ex2.getMessage());
                     throw new AuthenticationCompletionException(ex2);
                 }
             }

--- a/test-util/src/main/java/io/quarkiverse/kerberos/test/utils/KerberosKDCTestResource.java
+++ b/test-util/src/main/java/io/quarkiverse/kerberos/test/utils/KerberosKDCTestResource.java
@@ -148,7 +148,7 @@ public class KerberosKDCTestResource implements QuarkusTestResourceLifecycleMana
                         baos.write(second);
                         baos.write(substitute.toByteArray()); // Terminator never found.
                     }
-                    String toReplace = new String(substitute.toByteArray(), StandardCharsets.UTF_8);
+                    String toReplace = substitute.toString(StandardCharsets.UTF_8);
                     if (mappings.containsKey(toReplace)) {
                         baos.write(mappings.get(toReplace).getBytes());
                     } else {

--- a/test-util/src/main/java/io/quarkiverse/kerberos/test/utils/KerberosTestClient.java
+++ b/test-util/src/main/java/io/quarkiverse/kerberos/test/utils/KerberosTestClient.java
@@ -49,22 +49,23 @@ public class KerberosTestClient {
                             ValidatableResponse result = RestAssured.given()
                                     .header(HttpHeaderNames.AUTHORIZATION.toString(),
                                             NEGOTIATE + " " + Base64.getEncoder().encodeToString(token))
-                                    .get(path).then();
+                                    .get(path).then()
+                                    .log().all();
 
                             if (result.extract().statusCode() == 200) {
                                 return result;
                             } else if (result.extract().statusCode() == 401) {
                                 String header = result.extract().header(HttpHeaderNames.WWW_AUTHENTICATE.toString());
                                 if (header != null) {
-                                    if (header.length() > NEGOTIATE.toString().length() + 1) {
+                                    if (header.length() > NEGOTIATE.length() + 1) {
                                         // Negotiation continues
 
                                         byte[] headerBytes = header.getBytes(StandardCharsets.US_ASCII);
                                         // FlexBase64.decode() returns byte buffer, which can contain backend array of greater size.
                                         // when on such ByteBuffer is called array(), it returns the underlying byte array including the 0 bytes
-                                        // at the end, which makes the token invalid. => using Base64 mime decoder, which returnes directly properly sized byte[].
+                                        // at the end, which makes the token invalid. => using Base64 mime decoder, which returns directly properly sized byte[].
                                         token = Base64.getMimeDecoder().decode(
-                                                ArrayUtils.subarray(headerBytes, NEGOTIATE.toString().length() + 1,
+                                                ArrayUtils.subarray(headerBytes, NEGOTIATE.length() + 1,
                                                         headerBytes.length));
                                     } else {
                                         fail("Negotiation data has not been returned with WWW-Authenticate");

